python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -1234,54 +1269,89 @@ def get_subset_selection_for_context(
         dbt_resource_props = dbt_resource_props_by_output_name[output_name]
 
         # Explicitly select a dbt resource by its fully qualified name (FQN).
         # https://docs.getdbt.com/reference/node-selection/methods#the-file-or-fqn-method
         fqn_selector = f"fqn:{'.'.join(dbt_resource_props['fqn'])}"
 
-        selected_dbt_resources.append(fqn_selector)
+        selected_dbt_non_test_resources.append(fqn_selector)
 
+    selected_dbt_tests = []
     for _, check_name in context.selected_asset_check_keys:
         test_resource_props = dbt_resource_props_by_test_name[check_name]
 
         # Explicitly select a dbt resource by its fully qualified name (FQN).
         # https://docs.getdbt.com/reference/node-selection/methods#the-file-or-fqn-method
         fqn_selector = f"fqn:{'.'.join(test_resource_props['fqn'])}"
 
-        selected_dbt_resources.append(fqn_selector)
+        selected_dbt_tests.append(fqn_selector)
 
-    # Take the union of all the selected resources.
-    # https://docs.getdbt.com/reference/node-selection/set-operators#unions
-    union_selected_dbt_resources = ["--select"] + [" ".join(selected_dbt_resources)]
+    # if all asset checks for the subsetted assets are selected, then we can just select the
+    # assets and use indirect selection for the tests. We verify that
+    # 1. all the selected checks are for selected assets
+    # 2. no checks for selected assets are excluded
+    # This also means we'll run any singular tests.
+    selected_checks_target_selected_assets = all(
+        check_key.asset_key in context.selected_asset_keys
+        for check_key in context.selected_asset_check_keys
+    )
+    all_check_keys = {
+        check_spec.key for check_spec in assets_def.node_check_specs_by_output_name.values()
+    }
+    all_checks_included_for_selected_assets = all(
+        check_key.asset_key not in context.selected_asset_check_keys
+        for check_key in (all_check_keys - context.selected_asset_check_keys)
+    )
 
-    if context.is_subset:
+    if selected_checks_target_selected_assets and all_checks_included_for_selected_assets:
+        selected_dbt_resources = selected_dbt_non_test_resources + selected_dbt_tests
+        indirect_selection = None
         logger.info(
             "A dbt subsetted execution is being performed. Overriding default dbt selection"
-            f" arguments `{default_dbt_selection}` with arguments: `{union_selected_dbt_resources}`"
+            f" arguments `{default_dbt_selection}` with arguments: `{selected_dbt_resources}`. "
         )
+    # otherwise we select all assets and tests explicitly, and turn off indirect selection. This risks
+    # hitting the CLI argument length limit, but in the common scenarios that can be launched from the UI
+    # (all checks disabled, only one check and no assets) it's not a concern.
+    # Since we're setting DBT_INDIRECT_SELECTION=empty, we won't run any singular tests.
     else:
```

Comment:
we can get fancier in the future and use indirect selection in more cases. E.g. using select or exclude on certain tests, and indirect selection on the rest